core/emulate.c

```diff
@@ -1202,14 +1373,17 @@ em_status_t EMCALL em_emulate_insn(struct em_context_t *ctxt)
         rc = decode_operands(ctxt);
         if (rc != EM_CONTINUE)
             goto exit;
+        gpr_cache_flush(ctxt);
         goto restart;
     }
 
 done:
     rc = EM_CONTINUE;
     ctxt->finished = true;
     ctxt->ops->advance_rip(ctxt->vcpu, ctxt->len);
+    gpr_cache_flush(ctxt);
 
 exit:
+    gpr_cache_invalidate(ctxt);
```

Comment:
I suppose this is mainly to reset the GPR cache for the `EM_EXIT_MMIO` (re-entry to `em_emulate_insn()` for the same instruction) case, because the decoder already resets the GPR cache for every new instruction. But do you have an example that shows QEMU may indeed touch a guest GPR between `HAX_EXIT_FAST_MMIO` and the subsequent `HAX_VCPU_IOCTL_RUN`?